### DISCUSSION TITLE
JSON Configuration Languages

### DISCUSSION THREAD

**Discussion by matthewfeickert** [hep-statistics-standard/hep-statistics-standard#21] [Technical]
In the [March 24th meeting](https://indico.cern.ch/event/896473/) @lukasheinrich and @kratsg introduced several existing JSON configuration languages that would be interesting to follow up on how to adopt if possible, instead of trying to create a new one.

- [JSON-LD](https://json-ld.org/)
   - [JSON-LD 1.1 Framing](https://w3c.github.io/json-ld-framing/)
- [Jsonnet](https://jsonnet.org/)
- [Dhall](https://dhall-lang.org/)
- [CUE](https://cuelang.org/)

**Comment by lmoureaux** [comment]
For durability, I would advocate against using anything not on an standards track as part of the main specification (it's ok for convenience tools). As far as I can tell, this would rule out Jsonnet, Dhall, and CUE.

I would add [JSON Pointer](https://www.rfc-editor.org/rfc/rfc6901) (RFC 6901) to the list, since the current specification makes heavy use of object linking.

**Comment by cburgard** [comment]
@lmoureaux Thanks a lot for the suggestion, I think JSON Pointer sounds like an excellent idea, even though I'm not entirely sure of the use case yet. 

I think one of the main interests here was also to be able to use JSON files as "patches" to other JSON files by effectively "merging" them. This would allow extremely easy combination measurements, but also for example providing a bunch of different models for the same dataset or vice-versa.

Do people know if there's a good way to "include" JSON files from other JSON files?

**Comment by lmoureaux** [comment]
JSON pointer can be used to point to other elements in a JSON file. For instance, given this document:
```json
{ "a": { "value": 0, "error": 1 }}
```
The JSON Pointer `/a/value` points to 0 and `/a/error` points to 1. I think this would be useful to represent the edges of the evaluation graph.

> Do people know if there's a good way to "include" JSON files from other JSON files?

A common way (at least in JSON Schema and Swagger/OpenAPI) is to use [JSON Reference](https://datatracker.ietf.org/doc/html/draft-pbryan-zyp-json-ref-03) (expired draft), which can reference arbitrary nodes in external documents:
```json
{ "b": { "$ref": "https://example.com/example.json#json/reference/to/a" }}
```

**Comment by cburgard** [comment]
This sounds amazing, but can JSON reference also include whole lists of nodes? I'm envisioning a case where we want to do combinations without concatenating the nodes, and since all of our JSON files will be "flat" anyway, the only conceivable option is to merge them at top-level. Now, of course you could write a tool that does this, but I'm wondering if prescribing a standard way to merge serveral files into one would not be expedient.

**Comment by lmoureaux** [comment]
If you have `A.json`:
```json
{"x": [1, 2, 3], "y": [4, 5, 6]}
```
Then `A.json#x` references `[1, 2, 3]` and  `A.json#x/0` references `1`.
If you want to merge `x` and `y`, you need to specify every element:
```json
{"merged": [
  {"$ref": "A.json#x/0"},
  {"$ref": "A.json#x/1"},
  {"$ref": "A.json#x/2"},
  {"$ref": "A.json#y/0"},
  {"$ref": "A.json#y/1"},
  {"$ref": "A.json#y/2"}
]}
```

**Reply by kratsg** [reply]
This isn't totally true if you allow for GJSON/XPath syntax and check the type of object you're grabbing under a given pattern. For example, the above could be

```
{"merged": [
  {"$ref": "/x/*"},
  {"$ref": "/y/*"}
]}
```

**Reply by lmoureaux** [reply]
AFAIK this isn't supported by JSON Reference and would be a proprietary extension.

**Reply by kratsg** [reply]
Only if you're married to JSON for both format _and_ schema.

**Reply by lmoureaux** [reply]
I don't understand your comment. The advantage I see to using "standard" constructs is leveraging existing libraries instead of writing our own. If we deviate from what is supported, this point becomes moot. In fact I'd rather avoid things that look like existing standards but are slightly different, since this would be misleading.

**Reply by jpivarski** [reply]
Since you're defining a schema on top of basic JSON, anyway, you could introduce a rule in your schema that `"$ref"` means cross-reference. If it happens to coincide with a cross-reference syntax used in some other JSON extension, then that just makes it more familiar to some developers and easier to document.

Supporting cross-references, by itself, seems to me like a sufficiently minor thing that I wouldn't call it reimplementing other libraries, even if it is one of the features of those other libraries. There's a threshold of how much duplication is too much duplication, but this doesn't seem like too much to me.

That's how I would interpret @kratsg's suggestion. He can correct me if he meant it in a different spirit.

**Reply by kratsg** [reply]
That's what I meant, yes. GJSON/XPath are sufficiently widespread enough that supporting them is a minor caveat. The schema is defined using JSON, but `$ref` in JSON Schema is separate from the `$ref` we're trying to define in the actual probability models we want to use. Those probability models could be written in YAML or TOML, and we still would rely on JSON Schema to parse/validate the structure, but that doesn't mean the data itself is JSON.

**Comment by cburgard** [comment]
@lmoureaux Okay, that makes sense, but seems extremely cumbersome for our use-case...
In this case, we might want to go with something custom like:

```
{ 
  "external" : ["thefirstfile.json", "thesecondfile.json", "thethirdfile.json"]
  // this will merge the contents of thefirstfile.json, thesecondfile.json, thethirdfile.json in the following fashion: lists will be concatenated, and elements sharing the same name will be used with decreasing precedence.
}
```

But perhaps we are getting ahead of ourselves with this a little bit.

**Comment by lmoureaux** [comment]
I wasn't thinking of external references, I think most files should be self-contained (even if the prospect of referencing `https://cms.web.cern.ch/likelihoods/LumiRun2.hs3#likelihoods/0` from every CMS paper sounds interesting for the purpose of combinations).

The main application of JSON pointer I had in mind was for internal references. For instance, the current example for a likelihood is (simplified):
```json
{
  "name": "likelihood1",
  "distributions": [ "dist1", "dist2", "single_dimensional_dist_1" ],
  "data": [ "data1", "data2", 0 ]
}
```
With JSON pointer and the "struct" option of #2, this could be:
```json
"likelihood1": {
  "type": "likelihood",
  "distributions": [ "/dist1", "/dist2", "/single_dimensional_dist_1" ],
  "data": [ "/data1", "/data2", 0 ]
}
```
This may look very similar, but with the second option a JSON-Pointer-aware library could be used together with the schema to create the whole computational graph instead of doing it by hand.

**Comment by cburgard** [comment]
At first glance, that looks like a reasonable proposal. I do have a few questions though:

1. Elements of in the file are to some extent grouped by type, that is, all distributions live in `distributions`. Would we now end up in a situation where we would have 
```
"likelihood1": {
  "type": "likelihood",
  "distributions": [ "/distributions/dist1", "/distributions/dist2", "/distributions/single_dimensional_dist_1" ],
  "data": [ "/data/data1", "/data/data2", 0 ]
}
```
Are we okay with that, or do we feel this is too verbose / not friendly enough for human editing?

2. If we want to go with this approach, this would give a strong argument to use the dictionary-style-layout over the list-style-layout, no? (#2 )

**Comment by lmoureaux** [comment]
1. The syntax for arrays would be `/distributions/42` (where 42 is the array index). This is very much not user-friendly. For the key-value based structure, it's more verbose but unambiguous in cases where two objects have identical names, for instance `/distributions/variable1` and `/data/variable1`. It might also be more user-friendly when reading a file without knowing the format since one can immediately see which kind of object is being used.
2. Right.
